captum/optim/_utils/models.py

```diff
@@ -151,3 +155,72 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
             self.dilation,
             self.groups,
         )
+
+
+def pad_reflective_a4d(x: torch.Tensor, padding: List[int]) -> torch.Tensor:
+    """
+    Reflective padding for all 4 dimensions of an NCHW tensor
+    """
+
+    assert x.dim() == 4
+    assert len(padding) == 8
+
+    # Pad width
+    if padding[0] != 0:
+        x = torch.cat([x, x.flip([3])[..., 0 : padding[0]]], dim=3)
+    if padding[1] != 0:
+        x = torch.cat([x.flip([3])[..., -padding[1] :], x], dim=3)
+
+    # Pad height
+    if padding[2] != 0:
+        x = torch.cat([x, x.flip([2])[..., 0 : padding[2], :]], dim=2)
+    if padding[3] != 0:
+        x = torch.cat([x.flip([2])[..., -padding[3] :, :], x], dim=2)
+
+    # Pad channels
+    if padding[4] != 0:
+        x = torch.cat([x, x.flip([1])[:, 0 : padding[4]]], dim=1)
+    if padding[5] != 0:
+        x = torch.cat([x.flip([1])[:, -padding[5] :], x], dim=1)
+
+    # Pad batch
+    if padding[6] != 0:
+        x = torch.cat([x, x.flip([0])[0 : padding[6]]], dim=0)
+    if padding[7] != 0:
+        x = torch.cat([x.flip([0])[-padding[7] :], x], dim=0)
+    return x
+
+
+def collect_activations(
+    model,

```

Comment:
@ProGamerGov, Is this function and `ActivationCatcher` called anywhere except test cases ? Originally, as I checked out the code, it wasn't there

Reply:
It's used in the circuits file. The `ActivationCatcher` class exists to mimic how `InputOptimization` works so that it can reuse stuff, and so that notebook hook bug is prevented (hooks remain active or stop working for some reason). The `collect_activations` function makes it possible to use activation catch in one line of code.

Reply:
https://github.com/ProGamerGov/captum/blob/optim-wip/captum/optim/_utils/circuits.py

Reply:
The expanded weights function is use pretty heavily by the weight visualization notebook.

Reply:
@NarineK The expanded weights function is used pretty heavily by the weight visualization notebook, and the function itself required collecting activations in a way that the existing code didn't support easily.

Reply:
I see - thank you! I haven't started reviewing circuits yet. I'll take a look into it. Originally as I checked out the code couple days ago, I could see `ActivationCatcher` but `collect_activations ` isn't there. I need to rebase...